src/metabase/sync_database/analyze.clj

```diff
@@ -157,86 +59,99 @@
+          (nil? v)       (recur valid-count non-nil-count more)
+          :else          (let [valid? (and (string? v)
+                                           (u/is-url? v))]
+                           (recur (if valid? (inc valid-count) valid-count)
+                                  (inc non-nil-count)
+                                  more)))))
+
+(defn field-avg-length
+  "Default implementation of optional driver fn `field-avg-length` that calculates the average length in Clojure-land via `field-values-lazy-seq`."
+  [values]
+  (let [field-values       (filter identity values)
+        field-values-count (count field-values)]
+    (if (zero? field-values-count)
+      0
+      (int (math/round (/ (->> field-values
+                               (map str)
+                               (map count)
+                               (reduce +))
+                          field-values-count))))))
+
+(defn field-fingerprint [driver table field]
+  (let [values      (->> (driver/field-values-lazy-seq driver field)
+                         (take driver/max-sync-lazy-seq-results))
+        field-id    (:id field)
+        fingerprint {:base_type               (:base_type field)
+                     :is_pk                   (isa? (:special_type field) :type/PK)
+                     :is_fk                   (isa? (:special_type field) :type/FK)
+                     :cardinality             (count (distinct values))
+                     :field_percent_urls      (percent-valid-urls values)
+                     :field_percent_json      (if (values-are-valid-json? values) 100 0)
+                     :field_percent_email     (if (values-are-valid-emails? values) 100 0)
+                     :field_avg_length        (field-avg-length values)
+                     :field_id                field-id
+                     :table_id                (:id table)
+                     :name                    (:name field)
+                     :qualified_name          (field/qualified-name field)
+                     :visibility_type         (:visibility_type field)}]
+    (log/trace (u/format-color 'green "generated fingerprint for field: %s (%s):%s" field-id (:name field) fingerprint))
+    fingerprint))
+
+(defn- save-field-fingerprints!
+  "store a sequence of fingerprints"
+  [fingerprints]
+  (doseq [fingerprint fingerprints]
+    (let [fingerprint (-> fingerprint
+                          (update :base_type u/keyword->qualified-name)
+                          (update :visibility_type u/keyword->qualified-name))]
+         (log/trace (u/format-color 'cyan "saving fingerprint for field: %s (%s):%s"
+                      (:field_id fingerprint) (:name fingerprint) (keys fingerprint)))
+         (or (db/update! FieldFingerprint {:where [:= :field_id (:field_id fingerprint)]
+                                           :set fingerprint})
+             (db/insert! FieldFingerprint fingerprint)))))
```

Comment:
this is annoying but you could potentially hit a race condition here if two threads try to save a new fingerprint for the same field ID at the same. They could both try to hit `insert!` at the same time and one will fail because it's now a duplicate entry.